cmd/node.go

```diff
@@ -353,19 +565,36 @@ func (node *Node) listenToNodeMessages(exitChannels ...*chan struct{}) {
 	case <-node.internalExitChan:
 		break
 	case operation := <-node.nodeMessageChan:
-		if !node.IsRunning {
-			panic("Node.listenToNodeMessages: Node is currently not running, nodeMessageChan should've not been called!")
-		}
-		glog.Infof("Node.listenToNodeMessages: Stopping node")
-		node.Stop()
-		glog.Infof("Node.listenToNodeMessages: Finished stopping node")
 		switch operation {
+		case lib.NodeRestart:
+			// Using Mutex while accessing the Node status to avoid any race conditions
+			node.statusMutex.Lock()
+			if !node.IsRunning() {
+				panic("Node.listenToNodeMessages: Node is currently not running, nodeMessageChan should've not been called!")
+			}
+
+			glog.Infof("Node.listenToNodeMessages: Restarting node.")
+			glog.Infof("Node.listenToNodeMessages: Stopping node")
+			node.Stop()
+			node.statusMutex.Unlock()
+			glog.Infof("Node.listenToNodeMessages: Finished stopping node")
+
 		case lib.NodeErase:
+			glog.Infof("Node.listenToNodeMessages: Restarting node with a Database erase.")
+			glog.Infof("Node.listenToNodeMessages: Stopping node")
+			// cannot stop the node if the node is not already in RUNNING state.
+			// Stop the node and remove the data directory if the server is in RUNNING state.
+			node.statusMutex.Lock()
+			if node.IsRunning() {
+				node.Stop()
+			}
+			node.statusMutex.Unlock()
+			glog.Infof("Node.listenToNodeMessages: Finished stopping node")
+
 			if err := os.RemoveAll(node.Config.DataDirectory); err != nil {
 				glog.Fatal(lib.CLog(lib.Red, fmt.Sprintf("IMPORTANT: Problem removing the directory (%v), you "+
 					"should run `rm -rf %v` to delete it manually. Error: (%v)", node.Config.DataDirectory,
 					node.Config.DataDirectory, err)))
-				return
```

Comment:
why are we removing the `return` statement here?

Reply:
@lazynina `Fatal` would any way exit right?